lib/FrontendTool/FrontendTool.cpp

```diff
@@ -1623,20 +1623,37 @@ int swift::performFrontend(ArrayRef<const char *> Args,
   // arguments are generated by the driver, not directly by a user. The driver
   // is responsible for emitting diagnostics for its own errors. See SR-2683
   // for details.
-  //
-  // FIXME: Do we need one SerializedConsumer per primary? Owned by the
-  // SourceFile?
-  std::unique_ptr<DiagnosticConsumer> SerializedConsumer;
+  std::unique_ptr<DiagnosticConsumer> SerializedConsumerMultiplexer;
   {
-    const std::string &SerializedDiagnosticsPath =
-        Invocation.getSerializedDiagnosticsPathForAtMostOnePrimary();
-    if (!SerializedDiagnosticsPath.empty()) {
-      SerializedConsumer.reset(
+    SmallVector<FileSpecificDiagnosticConsumer::ConsumerPair, 4>
+      SerializedConsumers;
+    Invocation.getFrontendOptions().InputsAndOutputs
+        .forEachInputProducingSupplementaryOutput(
+           [&](const InputFile &Input) -> bool {
+      std::string SerializedDiagnosticsPath = Input.serializedDiagnosticsPath();
+      if (SerializedDiagnosticsPath.empty()) {
+        // If one input does not have a serialized diagnostic output path, none
+        // of them will.
```

Comment:
Is this checked somewhere with a diagnostic issued? Does it need to be?

Reply:
For noobs could add the following to the comment: "because serialized diagnostics are only used when Swift is invoked for Xcode, and in that case each primary input must communicate its diagnostics back to the IDE."

Reply:
Uh, hm. I was assuming this was guaranteed by the driver, not the IDE. I suppose there's no reason to build in that limitation here, though. I'll just return `false` instead of `true` to keep going.